src/main/java/org/truffleruby/builtins/BuiltinsCache.java

```diff
@@ -0,0 +1,202 @@
+            channel = FileChannel.open(file.toPath(), StandardOpenOption.READ);
+            reader = Channels.newReader(channel, "UTF-8");
+            lock = channel.lock(0, Long.MAX_VALUE, true);
+        }
+
+        public void close() throws IOException {
+            lock.release();
+            reader.close();
+            channel.close();
+        }
+
+        public Reader getReader() {
+            return reader;
+        }
+
+    }
+
+    public void cachedCoreMethodsAndPrimitives(List<List<? extends NodeFactory<? extends RubyNode>>> coreNodeFactories) {
+        Log.LOGGER.config("Regenerating builtins cache");
+
+        try (LockingFileWriter methodsWriter = new LockingFileWriter(coreMethodsCacheFile);
+                LockingFileWriter primitivesWriter = new LockingFileWriter(primitivesCacheFile);) {
+            final Writer methods = methodsWriter.getWriter();
+            final Writer primitives = primitivesWriter.getWriter();
+
+            for (List<? extends NodeFactory<? extends RubyNode>> nodeFactories : coreNodeFactories) {
+                for (NodeFactory<? extends RubyNode> nodeFactory : nodeFactories) {
+                    final Class<?> nodeClass = nodeFactory.getNodeClass();
+                    final CoreMethod method = nodeClass.getAnnotation(CoreMethod.class);
+                    Primitive primitiveAnnotation;
+                    if (method != null) {
+                        String moduleName = nodeClass.getEnclosingClass().getAnnotation(CoreClass.class).value();
+                        String type = method.isModuleFunction() ? "&" : method.onSingleton() || method.constructor() ? "." : "#";
+                        String visibility = method.visibility().name();
+                        methods.append(nodeFactory.getClass().getName()).append(SEPARATOR);
+                        methods.append(moduleName).append(SEPARATOR);
+                        methods.append(visibility).append(SEPARATOR);
+                        methods.append(type).append(SEPARATOR);
+                        final int rest = method.rest() ? 1 : 0;
+                        methods.append("" + method.required() + method.optional() + rest).append(SEPARATOR);
+                        final StringJoiner joiner = new StringJoiner(",");
+                        for (String name : method.names()) {
+                            joiner.add(name);
+                        }
+                        methods.append(joiner.toString());
+                        methods.append('\n');
+                    } else if ((primitiveAnnotation = nodeClass.getAnnotation(Primitive.class)) != null) {
+                        primitives.append(nodeFactory.getClass().getName()).append(SEPARATOR);
+                        primitives.append(primitiveAnnotation.name()).append('\n');
+                    }
+                }
+            }
+        } catch (IOException e) {
+            throw new JavaException(e);
+        }
+    }
```

Comment:
Yes, but how would you compute it at build time?
Run TruffleRuby in a special mode so it just generates the cache?
This kind of already happens since TruffleRuby is run to link C-exts.
But I also care about being able to run just by recompiling the `.jar` in Eclipse, and that does not re-trigger custom logic like C-exts.

An annotation processor would be one way. Then we probably don't even need the factories anymore. But it's also a lot more work.